cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/GitPATAuthenticationStrategy.java

```diff
@@ -30,15 +34,53 @@ public class GitPATAuthenticationStrategy implements AuthenticationStrategy {
   private static final String GITHUB_PAT_USERNAME = "oauth2";
 
   @Override
-  public CredentialsProvider getCredentialsProvider(SecureStore store,
-                                                    RepositoryConfig config,
-                                                    String namespaceId) throws AuthenticationConfigException {
-    try {
-      byte[] bytes = store.get(namespaceId, config.getAuth().getTokenName()).get();
-      String token = new String(bytes, StandardCharsets.UTF_8);
-      return new UsernamePasswordCredentialsProvider(GITHUB_PAT_USERNAME, token);
-    } catch (Exception e) {
-      throw new AuthenticationConfigException("Failed to get auth token from secure store", e);
+  public CredentialsProvider getCredentialsProvider(SecureStore store, RepositoryConfig config, String namespaceId) {
+    return new SecureStoreUserNamePasswordCredentialsProvider(store,
+                                                              GITHUB_PAT_USERNAME,
+                                                              config.getAuth().getTokenName(),
+                                                              namespaceId);
+  }
+
+  /**
+   * A class that wraps a {@link UsernamePasswordCredentialsProvider} but fetches password from a {@link SecureStore}.
+   */
+  private static class SecureStoreUserNamePasswordCredentialsProvider extends CredentialsProvider {
+    private final SecureStore secureStore;
+    private final String username;
+    private final String passwordTokenName;
+    private final String namespaceID;
+
+    SecureStoreUserNamePasswordCredentialsProvider(SecureStore secureStore,
+                                                   String username,
+                                                   String passwordTokenName,
+                                                   String namespaceID) {
+      this.secureStore = secureStore;
+      this.username = username;
+      this.passwordTokenName = passwordTokenName;
+      this.namespaceID = namespaceID;
+    }
+
+    @Override
+    public boolean isInteractive() {
+      return false;
+    }
+
+    @Override
+    public boolean supports(CredentialItem... credentialItems) {
+      return new UsernamePasswordCredentialsProvider("", "").supports(credentialItems);
+    }
+
+    @Override
+    public boolean get(URIish urIish, CredentialItem... credentialItems) throws UnsupportedCredentialItem {
+      try {
+        byte[] bytes = secureStore.get(namespaceID, passwordTokenName).get();
```

Comment:
though this entire implementation is messy as it doesn't really fit the jgit CredentialProvider interface. It will be very difficult to properly differentiate what is a securestore failure vs what is some other type of git failure.

Instead of this, my suggestion would be to add a refresh() method to the AuthenticationStrategy interface, which can declare some exceptions to throw. In refresh(), we lookup the required secure store values, and getCredentialsProvider just returns a plain UsernamePasswordCredentialsProvider. RepositoryManager refreshes when needed, and can catch exceptions there and know that it is an issue with secure store.

Reply:
Added a refresh() method.